### PR TITLE
Added serialization to the graph classes

### PR DESCRIPTION
**Pull request by svwilliams**
Added serialization to the graph classes

### CODE REVIEW

File: fuse_core/include/fuse_core/graph.h [hunk 1]
```diff
@@ -36,19 +36,102 @@
...
+  { \
+    return detail::type(); \
+  }
+
+/**
+* @brief Convenience function that creates the required pointer aliases, and type() method
+*
+* Usage:
+* @code{.cpp}
+* class Derived : public Constraint
```
Comment (efernandez) [inline fuse_core/include/fuse_core/graph.h]: `class Derived : public Graph`

File: fuse_core/include/fuse_core/graph.h [hunk 2]
```diff
@@ -36,19 +36,102 @@
...
+ * @brief Implements the type() member function using the suggested implementation
+ *
+ * Also creates a static detail::type() function that may be used without an object instance
+ *
+ * Usage:
+ * @code{.cpp}
+ * class Derived : public Graph
+ * {
+ * public:
+ *   FUSE_CONSTRAINT_TYPE_DEFINITION(Derived);
```
Comment (efernandez) [inline fuse_core/include/fuse_core/graph.h]: `FUSE_GRAPH_TYPE_DEFINITION(Derived);`

File: fuse_core/include/fuse_core/graph.h [hunk 3]
```diff
@@ -36,19 +36,102 @@
...
+
+/**
+* @brief Convenience function that creates the required pointer aliases, and type() method
+*
+* Usage:
+* @code{.cpp}
+* class Derived : public Constraint
+* {
+* public:
+*   FUSE_CONSTRAINT_DEFINITIONS(Derived);
```
Comment (efernandez) [inline fuse_core/include/fuse_core/graph.h]: `FUSE_GRAPH_DEFINITIONS(Derived);`

File: fuse_core/include/fuse_core/graph.h [hunk 4]
```diff
@@ -36,19 +36,102 @@
...
+*   // The rest of the derived constraint implementation
```
Comment (efernandez) [inline fuse_core/include/fuse_core/graph.h]: `// The rest of the derived graph implementation`

File: fuse_core/include/fuse_core/graph.h [hunk 5]
```diff
@@ -270,6 +361,73 @@ class Graph
...
+
+  /**
+   * @brief Serialize this Constraint into the provided binary archive
+   *
+   * This can/should be implemented as follows in all derived classes:
+   * @code{.cpp}
+   * archive << *this;
+   * @endcode
+   *
+   * @param[out] archive - The archive to serialize this constraint into
```
Comment (efernandez) [inline fuse_core/include/fuse_core/graph.h]: `constraint` -> `graph` (same below)

File: fuse_graphs/include/fuse_graphs/hash_graph.h [hunk 6]
```diff
@@ -330,8 +337,56 @@ class HashGraph : public fuse_core::Graph
...
+   * @param[in/out] archive - The archive object that holds the serialized class members
+   * @param[in] version - The version of the archive being read/written. Generally unused.
+   */
+  template<class Archive>
+  void serialize(Archive& archive, const unsigned int /* version */)
+  {
+    archive & boost::serialization::base_object<fuse_core::Graph>(*this);
+    archive & constraints_;
+    archive & constraints_by_variable_uuid_;
+    archive & problem_options_;
```
Comment (efernandez) [inline fuse_graphs/include/fuse_graphs/hash_graph.h:357]: I don't think we have to de/serialize the problem options. :thinking:

**Review by efernandez — APPROVED** [review]

File: fuse_graphs/include/fuse_graphs/hash_graph.h (hunk 6 above)
Comment (svwilliams) [reply fuse_graphs/include/fuse_graphs/hash_graph.h:357]: Agreed that it is probably overkill. But if you want to serialize a graph and later load it back and have it behave the same way, then it seems required. And its only a hand full of values, so it's not that painful.

Really what I'm toying with in my head moving a bunch of the graph functionality out into free functions. The graph would build the Problem object (where you could pass in Problem::Options at call time), then optimizing and computing covariances would be generic functions that operate on the Problem. But I have bigger fish to fry at the moment.

Comment (efernandez) [reply fuse_graphs/include/fuse_graphs/hash_graph.h:357]: Fair enough

**Review by ayrton04 — APPROVED** [review]

**Pull request merged**